### DISCUSSION TITLE
cant get gridpower from SMAplugin

### DISCUSSION THREAD

**Discussion by blema1966** [pkuehnel/TeslaSolarCharger#815] [Q&A] [answered]
Hello, first of all I would like to say that I am enthusiastic about the possibilities that arise with TeslaSolarCharger.

This is also the reason why I went into the big challenge as a beginner to install and I found it already  positive for me that I have the application running in the container. That both the connection to the car and the Teslamate dashboards work. However, my problem here is that I don't get the value for the grid power and I don't understand why. 

- In my setup there is an SMA inverter (192.168.188.114) and an SHM 2.0 (192.168.188.113). I have the SMA plugin running and the LOG shows that it has also joined the multicast group and is now listening on port 7191 for values, but so far in vain. I've been going in circles for many hours without success and even though I'm sure the problem is with me, I'd like to ask if anyone has any ideas as to how I can proceed here or where the problem could lie. Thank you in advance.

![Screenshot (1)](https://github.com/pkuehnel/TeslaSolarCharger/assets/133749728/fdb554cd-82ff-46a2-85d8-4ddacac50c77)

**Comment by pkuehnel** [comment, marked as answer]
On Windows, you need to run the SMA plugin without Docker. Just download the Windows binaries for the SMA Plugin in the Asset section from the [latest release](https://github.com/pkuehnel/TeslaSolarCharger/releases/latest), extract everything and start `Plugins.SmaEnergymeter.exe`.
Note: To let TSC access the values, you need to update the `apsettings.json`. Just add a `Kestrel` section below `"EnergyMeterMulticastAddress": "239.12.255.254"` The result should look like this:

```
{
  "Serilog": {
    "Using": [
      "Serilog.Sinks.Console"
    ],
    "MinimumLevel": {
      "Default": "Verbose",
      "Override": {
        "Microsoft": "Information",
        "System": "Error",
        "Microsoft.EntityFrameworkCore.Database.Command": "Information"
      }
    },
    "WriteTo": [
      {
        "Name": "Console",
        "Args": {
          "outputTemplate": "[{Timestamp:HH:mm:ss.fff} {Level:u3} {SourceContext}] {Message:lj}{NewLine}{Exception}"
        }
      }
    ],
    "Enrich": [
      "FromLogContext"
    ]
  },
  "AllowedHosts": "*",
  "EnergyMeterPort": 9522,
  "EnergyMeterMulticastAddress": "239.12.255.254",
  "Kestrel": {
    "Endpoints": {
      "Http": {
        "Url": "http://0.0.0.0:7192"
      }
    }
  }
}
```


**Original outdated answer:**
This is because the SMA plugin does not work on Docker on Windows, as it is running in a VM.

remove the SMA plugin from your docker compose file unzip [this file](https://pksoftwareservice-my.sharepoint.com/:u:/g/personal/pkuehnel_pk-softwareservice_de/EWKZZb6rK1lFiL_31aW3aFIBzCyFPIkZuUEJygpvqxzelg?e=vs5Fg0) and start `Plugins.SmaEnergymeter.exe` in it. You should then see something like this:
![image](https://github.com/pkuehnel/TeslaSolarCharger/assets/35361981/64e9fc39-26b4-4ccc-a642-5b4fdb9b2448)


If that is working update the Port  of the Grid URL in your TSC container from `7192` to `5000`, so the URL should look somewhat like that:

`http://192.168.178.35:5000/api/CurrentPower/GetPower`

**Reply by blema1966** [reply]
Many thanks for your very quik
![Screenshot 2023-05-17 173045](https://github.com/pkuehnel/TeslaSolarCharger/assets/133749728/960ba2ae-bb47-4750-ba40-5c7be2211a56)
![Screenshot 2023-05-17 173024](https://github.com/pkuehnel/TeslaSolarCharger/assets/133749728/5a7a8c1e-4e05-4e81-b240-76cd5e3b6761)
![Screenshot 2023-05-17 172807](https://github.com/pkuehnel/TeslaSolarCharger/assets/133749728/ab0d83a8-ef06-40b0-8710-6a88ee85924a)

**Reply by pkuehnel** [reply]
Set Result Type from `Json` to `Direct`

**Reply by blema1966** [reply]
I've already tried that but got the same result  


![Screenshot 2023-05-17 182141](https://github.com/pkuehnel/TeslaSolarCharger/assets/133749728/77381431-a21d-445c-aa8b-636de67d5db6)

**Reply by pkuehnel** [reply]
Hmm... what happens when you paste the URL in the browser?

**Reply by blema1966** [reply]
Than I will get ERR_CONNECTION_REFUSED from Browser. Once ein substitute host IP to "localhost" in Browser URL I ll get value.

Already switched off windows firewall on host, however get same result.



![Screenshot 2023-05-18 082056](https://github.com/pkuehnel/TeslaSolarCharger/assets/133749728/9a05050f-d57e-4b08-8613-ea1f268861cb)
![Screenshot 2023-05-18 082130](https://github.com/pkuehnel/TeslaSolarCharger/assets/133749728/04823736-0163-426e-a659-d1a4cb3bb267)
![Screenshot 2023-05-18 082215](https://github.com/pkuehnel/TeslaSolarCharger/assets/133749728/ef825829-a885-489e-9b7b-a6025119db9e)

**Reply by pkuehnel** [reply]
That's weird. Have you tried disabling your firewall of your laptop/pc? When you use localhost: The SMA Plugin should log something, when you refresh your browser (not quite sure if this really happens). Does it also happen when using the IP?

**Reply by blema1966** [reply]
Yes! Laptop firewall is off!. 

If I use localhost in the browser URL, I get new and correct values every time I refresh the browser.

If I use the IP address of the host, I only get the error message ERR_CONNECTION_REFUSED. Refreshing the browser gives the same result.

**Reply by pkuehnel** [reply]
Sorry, my bad. Use the following [file](https://pksoftwareservice-my.sharepoint.com/:f:/g/personal/pkuehnel_pk-softwareservice_de/Eq1QcdJGFotLowXVS4JOgbUB1kAamBdjAcAjq_3K7O82Jw?e=06NRE6). Only `appsettings.json` has changed. You probably get a Windows Defender notification on startup, which you need to allow.

**Reply by blema1966** [reply]
I'm so sorry to keep bothering you with this. Unfortunately, when I start the plugin, I'll get the following error message. however after 2 seconds plugin stoped.



![Screenshot 2023-05-18 185918](https://github.com/pkuehnel/TeslaSolarCharger/assets/133749728/2f29e58b-7a76-47cd-8ccc-f6324f77f1b5)

**Reply by blema1966** [reply]
I just removed HTTPS endpoint from appsettings.json file and now it is working :-) thank you so much!!!

**Reply by pkuehnel** [reply]
I added an issue https://github.com/pkuehnel/TeslaSolarCharger/issues/820 that, in future releases, all packages should also be published for Windows. But it should work as is for a long time. The last relevant change in the SMA Plugin was over a year ago, so no issues are expected when you update just the TSC part.

**Reply by pkuehnel** [reply]
@blema1966 From now on you find the Windows binaries on each [release ](https://github.com/pkuehnel/TeslaSolarCharger/releases/tag/v2.18.4) as `Plugins.SmaEnergymeter-vX.X.X-win-x64.tar.gz`. But as long as it is running, no update is needed (very likely nothing will change there). Note: The Kestrel settings are currently not correct in the releases, just keep your `appsettings.json` file. I will take care of that in a future release.
